gencpp.ml

```diff
@@ -1801,6 +1823,20 @@ and gen_expression ctx retval expression =
          gen_expression_list remaining
       ) in
 
+   (* this will add a cast if boxing / unboxing an objective-c type *)
+   let check_objc_unbox expression to_type =
+     if is_objc_type to_type && not (is_objc_type expression.etype) then
+        { expression with eexpr = TCast(expression,None); etype = to_type }
+     else
+        expression
+   in
+   let check_objc_box expression to_type =
+     if is_objc_type expression.etype && not (is_objc_type to_type) then
```

Comment:
This `not (is_objc_type to_type)` should probably be replaced by `is_dynamic_in_cpp`. I'll add more tests and make the change if needed